File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if(!head)return NULL;
        ListNode * temp=head;
        int count = 1;
        ListNode * leftn=NULL;
        ListNode * leftprev = NULL;
        ListNode * rightn = NULL;
        ListNode * rightnext = NULL;

        while(temp){
             if(count == left-1){
                leftprev=temp;
            }
            if(count == left){
                leftn=temp;
            }
            if(count == right){
                rightn=temp;
                rightnext=temp->next;
            }
            count++;
            temp=temp->next;
        }
        temp=leftn;
        stack<ListNode*> st;

        while(temp != rightn){
            st.push(temp);
            temp=temp->next;
        }
        st.push(rightn);

        ListNode* newhead=st.top();
         temp=st.top();
        st.pop();

        while(st.size()){
            temp->next=st.top();
            temp=temp->next;
            st.pop();
        }

        temp->next=rightnext;

        if(leftprev)
        leftprev->next=newhead;
        else head=newhead;

        return head;

    }
};

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head,ListNode end) {
        ListNode prev = end;
        ListNode curr = head;
        ListNode temp;

        while(curr!=end){
            temp = curr.next;
            curr.next=prev;
            prev=curr;
            curr=temp;
        }

        return prev;
    }

    public ListNode reverseBetween(ListNode head, int left, int right) {

        if(head==null)return null;
        ListNode leftnode=null;
        ListNode leftprev=null;
        ListNode rightnode=null;
        ListNode rightnext=null;

        ListNode temp = head;

        int i = 1;

        while(temp!=null){
            if(i == left-1){
                leftprev = temp;
            }
            if(i == left){
                leftnode = temp;
            }

            if(i == right){
                rightnode = temp;
                rightnext = temp.next;
                break;
            }
            i++;
            temp = temp.next;
        }

        ListNode mhead = reverseList(leftnode,rightnext);
        if(leftprev==null)head = mhead;
        else leftprev.next = mhead;

        leftnode.next=rightnext;
        return head;

    }
}